Mathlib/Algebra/Category/ModuleCat/TensorHomAdjunction.lean

```diff
@@ -0,0 +1,213 @@
+/-
+Copyright (c) 2023 Jujian Zhang. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Jujian Zhang
+-/
+import Mathlib.LinearAlgebra.TensorProduct
+import Mathlib.Algebra.Category.ModuleCat.Basic
+import Mathlib.CategoryTheory.Adjunction.Limits
+import Mathlib.LinearAlgebra.Finsupp
+
+/-!
+# Tensor-Hom adjunction
+Consider two commutative rings `R` and `S` and `X` an `(R, S)`-bimodule.
+Consider the tensor functor `(X ⊗[R] .)` from the category of `R`-module to the category of
+`S`-module and the hom functor `X →ₗ[S] .` from the category of `S`-module to the category of
+`R`-module. They form an adjunction. In particular we have that
+```
+Hom_S(X⊗[R]Y, Z) ≃ Hom_R(Y, Hom_S(X, Z))
+```
+
+## Implementation notes
+
+1. Order of arguments
+```
+Hom_S(Y⊗[R]X, Z) ≃ Hom_R(Y, Hom_S(X, Z))
+```
+is perhaps more natural to work with because all the argument have the same order.
+But currently mathlib4 does not know `Y⊗[R] X` is an `S`-module as well.
+
+2. Why not use the [`Tower` file](Mathlib/LinearAlgebra/TensorProduct/Tower.lean`)
+
+In our setting `X` is an `(R, S)` bimodule and `Y` an `R`-module and `Z` an `S`-module
+so to use the `Tower` file, we need `S` to be an `R`-algebra which is a luxury we do not have.
+But note that in `Tower` file, `curry` and `uncurry` are both tri-linear maps. So `Tower` file
+allows interplay of 3 rings which is not allowed in this file.
+
+-/
+
+universe u u' v
+
+open TensorProduct CategoryTheory
+
+variable (R : Type u) (S : Type u') (X : Type v)
+variable [CommRing R] [Ring S]
+variable [AddCommGroup X] [Module R X] [Module Sᵐᵒᵖ X] [SMulCommClass R Sᵐᵒᵖ X]
+
+namespace ModuleCat
+
+/--
+Let `X` be an `(R,S)`-bimodule. Then `(X ⊗[R] .)` is a functor from the category of `R`-modules
+to the category of `S`-modules.
+-/
+@[simps!]
+noncomputable def tensorFunctor : ModuleCat.{v} R ⥤ ModuleCat.{v} Sᵐᵒᵖ where
+  obj Y := ModuleCat.of Sᵐᵒᵖ <| X ⊗[R] Y
+  map {Y Y'} l :=
+  { __ := l.lTensor _
```

Comment:
I think the approach here is backwards; you should do everything universe-polymorphically *outside* category theory, then recover the category-theoretic version as a special case.

Reply:
In commite [63ec5b0](https://github.com/leanprover-community/mathlib4/pull/8495/commits/63ec5b0710f344c77c1bf2bb16fb0c60f39cfe85) I have moved `(un)curry` outside and recovered the category-theoretic version as a special case. Can you check if this is what you meant?